### PR TITLE
Shooter high level

### PR DESCRIPTION
**Pull request by rakrakon**
Subsystem class 
IO Interface class
Subsystem Commands

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    public boolean atSetpoint() {
+        return io.atSetpoint();
+    }
+
+    /** Updates the state of the shooter. */
+    @Override
+    public void periodic() {
+        io.setVelocity(inputs.velocitySetpoint);
+
+        io.updateInputs();
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: Passing the setpoint to the io should happen after updating the inputs.

**Review by vichik123 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/shooter/ShooterIO.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    ShooterInputsAutoLogged inputs = new ShooterInputsAutoLogged();
+
+    /**
+     * Set the position of the hood
+     *
+     * @param velocity The position of the subsystem.
+     */
+    void setVelocity(MutableMeasure<Velocity<Angle>> velocity);
+
+    boolean atSetpoint();
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/ShooterIO.java]: This is something that should be handled in the subsystem, it'll be the same for sim and real.

File: src/main/java/frc/robot/subsystems/shooter/ShooterIO.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import static edu.wpi.first.units.Units.*;
+
+import edu.wpi.first.units.*;
+import org.littletonrobotics.junction.AutoLog;
+
+public interface ShooterIO {
+    ShooterInputsAutoLogged inputs = new ShooterInputsAutoLogged();
+
+    /**
+     * Set the position of the hood
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/ShooterIO.java]:
```suggestion
     * Sets the position of the hood.
```

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    public static void initialize(ShooterIO io) {
+        INSTANCE = new Shooter(io);
+    }
+
+    /**
+     * Sets the velocity of the shooter.
+     *
+     * @param velocity The velocity of the Shooter to set.
+     */
+    public Command setVelocity(Supplier<Double> velocity) {
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: Why not receive a `Supplier<Measure>`/`Supplier<MutableMeasure>`?

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 1 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: e5a5682

**Review by katzuv — CHANGES_REQUESTED** [review]
Off to a great start! 🎶

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 4 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: 57ba347

File: src/main/java/frc/robot/subsystems/shooter/ShooterIO.java (hunk 2 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/ShooterIO.java]: 7a4bf78
5c91961

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    public MutableMeasure<Velocity<Angle>> getVelocity() {
+        return inputs.velocity;
+    }
+
+    public boolean atSetpoint() {
+        return Utils.epsilonEquals(
+                inputs.velocity.in(Units.RotationsPerSecond),
+                inputs.velocitySetpoint.in(Units.RotationsPerSecond),
+                ShooterConstants.atSetpointTolerance);
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: Should use [`isNear()`](https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/units/Measure.html#isNear(edu.wpi.first.units.Measure,double)) here.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    /** Updates the state of the shooter. */
+    @Override
+    public void periodic() {
+        io.updateInputs();
+        Logger.processInputs("Shooter", inputs);
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]:
```suggestion
        Logger.processInputs(this.getClass().getSimpleName();, inputs);
```
That way if the name of the class changes, you won't need to change it here.

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 5 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: c749683

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 6 above)
Comment (rakrakon) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: 97f1ed0

File: src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java [hunk 7]
```diff
@@ -0,0 +1,5 @@
+package frc.robot.subsystems.shooter;
+
+public class ShooterConstants {
+    public static double atSetpointTolerance = 0.05; // Percentages
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java]: This is probably too high, a 5% tolerance means that with a setpoint of 1000RPM, we'll be fine with 1050RPM which might be a significant error. We will probably achieve 0.5% tolerance.
Also, "setpoint tolerance" is enough, no need for the "at". Note also that constants should be written in the `ALL_CAPS` format.

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java (hunk 7 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java]: Skibidi yes yes

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+                inputs.velocitySetpoint, ShooterConstants.atSetpointTolerance);
+    }
+
+    /** Updates the state of the shooter. */
+    @Override
+    public void periodic() {
+        io.updateInputs();
+        Logger.processInputs(this.getClass().getSimpleName(), inputs);
+
+        io.setVelocity(inputs.velocitySetpoint);
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: This should be handled in the subsystem's default command, not here.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    public boolean atSetpoint() {
+        return inputs.velocity.isNear(
+                inputs.velocitySetpoint, ShooterConstants.atSetpointTolerance);
+    }
+
+    /** Updates the state of the shooter. */
+    @Override
+    public void periodic() {
+        io.updateInputs();
+        Logger.processInputs(this.getClass().getSimpleName(), inputs);
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: Just realized `this.getClass().getSimpleName()` might not be efficient enough to be called every loop -- it should be moved into a constant.

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java (hunk 7 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java]: Also, I'm not sure the error should be percentage but rather in rps. The farther we are the more consistent the velocity would need to be, hence the error would need to be smaller. Using percentages means the opposite.

Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java]: That's how `isNear()` works though. I think we can see in testing whether the tolerance is proportional or constant. That means, if the setpoint is 100 then 101 might be fine, and if the setpoint is 1000 then 1010 might be fine too.

Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java]: Ok, we'll leave it for now.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 8 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: f26c70cd1cb92b9b81255417d747d8f320fa1430

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 9 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: 9749c7eb86a582284162b9bb0e63e533ea3b4542

Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: Perhaps "subsystem name" is better (e.g. in subsystems where `Mechanism2d` is used you need the subsystem name too)

Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: 8facfddafc5008747abf639b28be4be14a1d7020

Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: Sorry, last one… should be all caps, no? It's a constant

Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: 96066e7ef77b62bac6093a9cb4f074abc5080067

**Review by katzuv — CHANGES_REQUESTED** [review]
Missing [`Command.withName()`](https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/wpilibj2/command/Command.html#withName(java.lang.String)) decorators everywhere (should be written like normal text and not in `camelCase`.

**Pull request merged**